### ISSUE TITLE
Разделение проекта на шаблон и реализацию

### ISSUE THREAD

**Issue by klimentyev** [latex-g7-32/latex-g7-32#50]
Никто для себя при работе с проектом не разделял репозиторий на 2 репозитория: стили со сборочной системой и непосредственно исходники .tex, рисунки, листинги для конкретного документа?
Наверняка каждый из нас создает не единственный документ, используя этот стиль. Логично было бы иметь репозиторий со стилем, постоянно актуализировать его, что-то добавлять/изменять самому, при этом не затрагивая документ. В остальных репозиториях только файл в для документов.
Наверное, через символьные ссылки можно организовать. Пока не погружаться в эту работу. Вдруг у кого-нибудь уже были подобные мысли.

**Comment by einhander** [comment]
Я всегда использую для себя только стили, без обвязки в виде системы сборок и тп. Мысль разделить репозиторий есть уже давно, даже есть такой issue #17. Собственно все упиралось в отсутствие человеческих ресурсов. Сейчас люди заинтересованные в дальнейшем развитии появились (во всяком случае я на это надеюсь) поэтому главный вопрос как грамотнее поступить.
Любое перемещение стилевых файлов вызовет переделывание систем сборки.
Если их авторы не против, то можно сделать. @ZyX-I и @dvarubla что скажете?

P.S. но в любом случае  сперва нужно будет подождать готовности #47 пофиксить #49 и релизнуться, а то уже добавлений много;)

**Comment by einhander** [comment]
>Наверное, через символьные ссылки можно организовать.

@klimentyev не уверен что будет работать на win системах.

**Comment by klimentyev** [comment]
@einhander с Vista на win появилась поддержка символьных ссылок: https://ss64.com/nt/mklink.html
Такое чувство, что как-то на Win7 работал с проектом, который сделан на Linux с символьными ссылками. А так да, требует проверки.

**Comment by dvarubla** [comment]
Попробовал так сделать с помощью моего скрипта. Если в одном месте директория со стилями, .latexmkmodrc, build.sh, а в другом — директории tex, src, graphics, то  это несложно сделать и без символьных ссылок. По крайней мере, на Linux. Можно также подключить дополнительный конфигурационный файл с настройками для текущего проекта.

**Comment by ZyX-I** [comment]
Стиль — имеет смысл выделить и отправить на CTAN, если кто‐то документацию к нему напишет. Систему сборки — ни в коем случае, это неотъемлемая часть любого проекта. Когда вы создаёте свой проект на основе шаблона, вам нужно быть уверенным, что он не сломается из‐за действия третьих лиц. Ещё её иногда нужно подпиливать под себя. И в любом случае, система сборки в отдельно репозитории затрудняет развёртывание и автоматизацию сборки: попробуйте настроить travis или что ещё так, чтобы изменения в репозитории myuser/latex-g7-32-build (форк системы сборки, мы ведь не хотим, чтобы третьи лица что‐то сломали?) вызывали пересборку репозитория myuser/dip (собственно диплом, форк шаблона), при этом изменения в myuser/dip также должны вызывать его пересборку.

**Comment by ZyX-I** [comment]
Уточнение про utils:

- utils/pdfcrop можно было бы отправлять куда подальше, к системе сборки он имеет не большее отношение чем, к примеру, тот же dia. Т.е. для сборки нужен, но частью системы не являются — зависимость. Другое дело, что этот скрипт три года не менялся. И ещё, в моей системе есть pdfcrop, который есть часть пакета app-text/texlive-core. При чём: `diff =pdfcrop utils/pdfcrop` показывает ноль отличий. Так что просто выкинуть, я абсолютно уверен в том, что наличие pdfcrop там это не частная инициатива авторов пакета texlive-core в Gentoo.
- utils/texdepend нужен для Makefile, прямая часть системы, которую, впрочем, лучше бы вообще удалить из‐за проблем с minted и макросами.
- utils/install.py лучше держать в lyx (или хотя бы назвать понятнее).

**Comment by einhander** [comment]
@ZyX-I про utils/pdfcrop  полностью согласен, здесь оно не нужно. Для удаления поменять только в Make файле строчку.
install.py это установщик стиля в локальный texmf пользователя, не проверял как он работает, но нужен не только для lyx.
Сборку с помощью make предлагаю все же оставить, до тех пор пока она не станет полностью не работоспособной. В документации поменять порядок использования систем сборки.

Мое мнение - перенести стилевые файлы и документацию (будущую) в отдельный репозиторий и его подключить к текущему как submodule.  Симлинки сгодятся только как временное решение.

И кстати, [**задирает голову в верх**] @rominf и @imz мне нужно быть членом организации чтобы создать в ней новый репозиторий со стилевыми файлами.

**Comment by ZyX-I** [comment]
Я предлагал удалить texdepend, а не Makefile. Вроде Makefile вполне себе позволяют использовать шаблоны, а не имена файлов. Некоторые диалекты точно.

install.py я ни разу нигде не запускал, и так работает. Пока стиль находится там, где он находится, он нужен только для lyx. Если он будет в отдельном репозитории, то нужно смотреть как это делается в других стилях с CTAN. Вроде нужно просто расположить файлы в zip архиве как для распаковки в ~/texmf.

Подмодуль — не уверен. Не логичнее ли просить воспользоваться `tlmgr` для установки с CTAN?

**Comment by ZyX-I** [comment]
В docker, кстати, логично стиль просто запихнуть как один из установочных шагов (лучше, думаю, не в имеющемся RUN, а в следующем — чтобы можно было не пересобирать контейнер по часу после обновления стиля).

Кстати, я тут недавно узнал, что docker под Windows есть. Т.е. наш контейнер — это готовый способ сборки под Windows, не требующий установки кучи зависимостей — только сам docker, ну и контейнер придётся либо собрать, либо загрузить; и то, и то — одной командой, только разными (в README есть команда для сборки, т.к. «официально» контейнер не загружен).

**Comment by einhander** [comment]
Согласен, релизную версию стиля удобно тянуть с CTAN при помощи  `tlmgr`.
А что делать если хочется протестировать новую  версию стиля?

Это не docker в windows есть, они туда целый линкус туда запихали)) Можно в Readme про это написать.

**Comment by ZyX-I** [comment]
Разработчик стиля сам разберётся. Не релизную версию я бы как раз на месте пользователей тянуть не стал, особенно по‐умолчанию.

**Comment by einhander** [comment]
Что то я запутался, что вы предлагаете?

**Comment by ZyX-I** [comment]
Просто выделить стиль в отдельный репозиторий, отправить на CTAN и описать варианты установки как оттуда, так и из CTAN.

**Comment by einhander** [comment]
@ZyX-I Поддерживаю, так и поступим.
Небольшой роуд мап:
-  Добиваем вторую систему сборки.
-  Делаем релиз того что есть.
-  Выделяем стиль в отдельный репозиторий.
-  Пишем документацию.
-  Отправляем на CTAN

**Comment by LSinev** [comment]
Не выяснится ли случайно, что проектируемый для CTAN стиль или `documentclass` не будет сильно отличаться от того (тех), что идут с [пакетом `disser`](http://ctan.org/pkg/disser)? https://github.com/polariton/disser
https://github.com/polariton/disser/blob/master/src/gost732.dtx

**Comment by klimentyev** [comment]
@LSinev Насколько я понимаю, что тот стиль поддерживает только LaTeX. Отличие в виде поддержки XeLaTeX - это большое преимущество :)

**Comment by LSinev** [comment]
Ну если он три компилятора не поддерживает, то его можно доработать. Тем более, что и тут не всегда XeLaTeX был.

**Comment by einhander** [comment]
@LSinev текущий стиль сильно отличается от того что поставляется с пакетом disser. Первоначальный автор стиля Алексей Томин. Автор пакета disser  Станислав Кручинин, так что это разные стили. И в целом они выполняют разную функцию. Выпускные работы и диссертации требуют оформления по ГОСТ 7.32, но содержание разделов у них своё. Наш стиль реализует в том числе и содержание отчёта по НИР требуемое ГОСТ.

**Comment by LSinev** [comment]
Диссертации не по ГОСТ 7.32 делаются, а по ГОСТ Р 7.0.11.
Там уже человек регулярно наполняет CTAN и можно было бы воспользоваться его опытом. Сделать необходимые изменения до полного соответствия ГОСТ 7.32 в уже имеющемся там стиле.

**Comment by einhander** [comment]
>Диссертации не по ГОСТ 7.32 делаются, а по ГОСТ Р 7.0.11.

Вы ГОСТ Р 7.0.11. читали?  
>Сделать необходимые изменения до полного соответствия ГОСТ 7.32 в уже имеющемся там стиле.

Предложите ему PR я думаю он не откажется.

>Там уже человек регулярно наполняет CTAN и можно было бы воспользоваться его опытом. 

Спросить опытного человека всегда полезно.

**Comment by LSinev** [comment]
> Вы ГОСТ Р 7.0.11. читали?

[Читал](https://github.com/AndreyAkinshin/Russian-Phd-LaTeX-Dissertation-Template/blob/master/Documents/GOST%20R%207.0.11-2011.pdf)

**Comment by einhander** [comment]
Значит видели ссылку на ГОСТ 2.105 который регламентирует оформление  для обоих стандартов.

**Comment by LSinev** [comment]
Видел.
Это явно не то же самое, что:
> Выпускные работы и диссертации требуют оформления по ГОСТ 7.32

**Comment by einhander** [comment]
Вот вам ссылка на требования к выпускным работам и диссертациям [тыц](http://standard.tpu.ru/docs/standorg/ВКР_ориг1.htm), к тому же до 2012 госта не существовало, и вак требовал оформления по 7.32.

**Comment by LSinev** [comment]
Мы явно не студенты ТПУ. Сейчас уже пять лет как не 2012 год. Зато хотя бы понятно, чего это многие  вставляют УДК на титульник диссертации (по привычке/требованию научника или ученого совета, которые много диссертаций делали по 7.32).

**Comment by einhander** [comment]
ТПУ это как лучший пример оформления требований к выпускным работам, это хотя бы стандарт предприятия, что по текущим временам уже круто.
В остальных все много хуже, но большинство все так же требует 7.32 что на бакалавра, что на магистра. И непременно редактор MS Word и шрифт  Times New  Roman 14 пт.
Хотя кое где встречал, что требуют ЕСКД... в Ворде))

**Comment by dvarubla** [comment]
Мне кажется, @LSinev в чём-то прав. Попробовал скомпилировать бакалаврскую работу из репозитория, изменив стиль на gost732 — очень похоже на то, что у нас. Есть, конечно, мелкие недочёты (вроде полужирных заголовков, недостаточного отступа слева, разного размера шрифта, странной таблицы) но все равно, думаю, будет странно наличие практически одинаковых стилей в CTAN.

**Comment by einhander** [comment]
В CTAN есть дублирующие стили, пакет gost и vak, eskd и eskdx, это то что я быстро нашел, не говоря уже про таблицы - tabu, longtable, supertabular и др. В disser полно мелких недочетов, и все что у нас делается автоматически командами там нужно делать ручками. И к тому же у стилей различная направленность, disser про кандидатские диссертации (и уже потом все остальное), наш про отчет по НИР.

**Comment by klimentyev** [comment]
Решил попробовать использовать редактор Kile. Не подскажете, как в нем настроить сборку, используя Makefile? Версия Kile: 2.1.3, в настройках в меню Build пытался задавать pipe вида `cd <путь к каталогу> && make`, но собрать так и не получилось. Может быть, что-то упустил в параметрах. К сожалению, в сети не удалось найти примера, где было бы подробное описание.

**Comment by afaikiac** [comment]
В репозитории сейчас много устаревших сборок, которые не работают. Я уверен, что для многих пользователей это многообразие выглядит пугающе. При этом стили в репозитории лучше всего соответствуют гост732 и используют xelatex.

Я хотел бы предложить следующие изменения:

- Оставить только сборку с использованием cmake. CMakeLists.txt написан очень хорошо и понятно. И сборка, кажется, кроссплатформенная.

- Оставить Dockerfile для тех, кто не хочет устанавливать зависимости локально. На Windows  тоже есть docker.

- Выделить стили в отдельный репозиторий.

- Временно подгружать стили просто как сабмодуль и доработать `CMakeLists.txt` для этого.

- Полностью переписать README. Оставить только базовую информацию о том, что это за репозиторий и как его запустить.

Затем можно и дефолтную ветку поменять и почистить issues.

Что вы думаете об этих изменениях, @einhander?

**Comment by rominf** [comment]
Я не @einhander (я тот, кто перенёс чужие разработки на GitHub), но в целом одобряю. Readme и многообразие сборок действительно могут отпугнуть. Я бы ещё оставил LyX, если он в нормальном состоянии (давно не смотрел), поскольку у LyX меньше порог входа, чем у LaTeX, но это не первостепенно, не думаю, что у него очень много пользователей.

**Comment by einhander** [comment]
@afaikiac 
> В репозитории сейчас много устаревших сборок, которые не работают. 

Я бы не был столь категоричен:) В linux cmake и latexmk-mod точно работают, make вроде бы тоже. Что в windows не знаю.
> Я уверен, что для многих пользователей это многообразие выглядит пугающе.

Согласен.
>При этом стили в репозитории лучше всего соответствуют гост732 и используют xelatex.

Можете считать меня ретроградом, но без pdflatex никуда, до не давнего времени был стабилен как скала, с переходом на latex3 могут быть неожиданности. С xelatex иногда есть проблемы, например пакеты soul(utf8).

@rominf 
>Я бы ещё оставил LyX, если он в нормальном состоянии (давно не смотрел)

В ненормальном, тут есть issue по этому поводу. 

А так идея хорошая, правильная и нужная. Мое мнение - поддержка xelatex и pdflatex. С остальным согласен.

**Comment by rominf** [comment]
@einhander Действительно, не заметил. Ну тогда можно LyX выкидывать. Если кто-то захочет его вернуть, то возьмёт из истории и допилит.

**Comment by einhander** [comment]
Тут еще можно выкинуть Inkscape из примера, он используется только для конвертации изображений. Смысла в нем тут нет, если цель показать, что можно так делать, то проще заменить на что-то попроще.

**Comment by afaikiac** [comment]
> Тут еще можно выкинуть Inkscape из примера

Наверно подразумевается, что создание и редактирование svg происходит в Inkscape, поэтому он есть в системе. Но можно подумать и о замене. Я бы вообще сделал все эти доп обработчики условными.

> В linux cmake и latexmk-mod точно работают, make вроде бы тоже.

В любом случае мейнтейнить несколько сборок трудозатратно. Поэтому оставить лучше что-то одно. Если cmake хорошо собирает и pdflatex, я за то, чтобы оставить только его. У него и синтаксис простой и документация приятная. Любой разберется и доработает под себя.

Еще я бы добавил минимальный конфиг для latexmk, потому что vimtex c ним хорошо работает. Думаю большинство ограничивается  вставкой простых картинок, и все эти обработчики из cmake не нужны, но приятная фишечка.

**Comment by afaikiac** [comment]
Я правильно понимаю, что выделять в отдельный репозиторий стоит только:

G2-105.sty
G7-32.cls
G7-32.sty

И в последующем для их генерации надо будет написать .dtx файлы.

Остальное  --- это локальные стили, которые каждый под себя может настроить?

**Comment by einhander** [comment]
* GostBase.clo - часть стиля
* gosttitleGost7-32.sty - Стиль титульной страницы по 7-32 (часть стиля)
* gosttitleGostRV15-110.sty - Поддержка закрытого ГОСТ РВ 15-110, вроде бы даже полная

**Comment by afaikiac** [comment]
Сейчас основная проблема репозитория: чтобы сбилдить шаблон, нужно использовать скрипты.
Можно уйти от этого и разделить все системы сборки на ветки или репозитории.

Например:
`G7-32` репозиторий для стилей
`latex-g7-32-cmake` ветка или репозиторий для cmake, и тд
`latex-g7-32` будет использовать только `latexmk`, и его без труда можно будет и в `overleaf` залить.

Таким образом и скрипты сборки можно сохранить и сделать репозиторий дружелюбным.

**Comment by einhander** [comment]
В целом поддерживаю такое разделение.  Шрифты, я бы не выкидывал, шаблон на них настроен, CM совсем не похож на Times который рекомендуют в ГОСТ. Можно положить в отдельный репозиторий.

**Comment by afaikiac** [comment]
Сделал ветку, которая работает с overleaf и собирается с помощью latexmk https://github.com/afaikiac/latex-g7-32. Стили подгружаются как сабмодуль из другой ветки, изменения добавлены с коммита `591b305`. Докер весит ~800M.

В репозитории можно создать, например, ветки `G7-32` и `7.32-2017-overleaf`. И еще аккаунт в docker hub, чтобы туда загрузить официальный контейнер в котором "точно соберется". С помощью него можно будет легко и быстро тестировать все изменения стилей.

> Шрифты, я бы не выкидывал, шаблон на них настроен

Добавил официальные ссылки на шрифты в README :)

**Comment by afaikiac** [comment]
@einhander вижу Вы смерджили ветку со стилем в #125. А в планах есть мерджить ветку под оверлиф с github actions? Она сейчас немного ориентирована под мои нужды, но могу сделать более общий вариант и добавить несколько фишечек.

**Comment by einhander** [comment]
>А в планах есть мерджить ветку под оверлиф с github actions?

Да, в планах такое есть. Если есть желание сделать более общий вариант, могу только поприветствовать такое решение.

> Вы смерджили ветку со стилем в #125 

Есть желание выделить ее в отдельный репозиторий, что наверное будет более правильным. А этот репозиторий оставить как референсную реализацию и пример использования.

**Comment by afaikiac** [comment]
> Есть желание выделить ее в отдельный репозиторий

Мне кажется лучше это сразу сделать, иначе сложно поддерживать, когда 2 разных проекта в 1 репозитории.

Я могу потом сразу сделать pr, чтобы стили подгружались как сабмодуль в основную ветку с шаблоном. И осталась поддержка cmake. 

И все изменения по стилям вносить в новый репозиторий.

**Comment by einhander** [comment]
>Мне кажется лучше это сразу сделать, иначе сложно поддерживать, когда 2 разных проекта в 1 репозитории.

Готово.
